File: Python/Linked_List/2.两数相加.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
	def addTwoNumbers(self, l1, l2):
		"""
		:type l1: ListNode
		:type l2: ListNode
		:rtype: ListNode
		"""
		head = ListNode(0)
		node = head
		ADD = 0
		while l1 != None or l2 != None or ADD != 0:
			if l1 != None:
				ADD += l1.val
				l1 = l1.next
			if l2 != None:
				ADD += l2.val
				l2 = l2.next

			node.next = ListNode(ADD%10)
			if ADD >= 10:
				ADD = int(ADD/10)

		return head.next

File: Go/ArrayList/2-两数相加.go
package ArrayList

import "fmt"

type ListNode struct {
	Val int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	//var ans *ListNode	// todo: * 和 & 在声明变量时的区别
	ans := ListNode{
		Val:  0,
		Next: nil,
	}
	var head ListNode
	head = ans
	f := 0
	for l1 != nil && l2 != nil {
		tmp := &ListNode{}
		v := l1.Val + l2.Val + f
		if v >= 10 {
			f = 1
			v = v - 10
		} else {
			f = 0
		}
		tmp.Val = v
		ans.Next = tmp
		ans = *ans.Next
		l1 = l1.Next
		l2 = l2.Next
	}
	head = *head.Next
	return &head
}

func pprint(l *ListNode) {
	n := l
	for n != nil {
		fmt.Print(n.Val)
		n = n.Next
	}
}